lightning/src/ln/channelmanager.rs

```diff
@@ -106,6 +106,19 @@ impl Readable for InFlightHtlcs {
 	}
 }
 
+/// A trait defining behavior for routing an payment.
+pub trait Router {
+	/// Finds a [`Route`] between `payer` and `payee` for a payment with the given values.
+	fn find_route(
+		&self, payer: &PublicKey, route_params: &RouteParameters,
+		first_hops: Option<&[&ChannelDetails]>, inflight_htlcs: InFlightHtlcs
+	) -> Result<Route, LightningError>;
+	/// Lets the router know that payment through a specific path has failed.
+	fn notify_payment_path_failed(&self, path: &[&RouteHop], short_channel_id: u64);
+	/// Lets the router know that payment through a specific path was successful.
+	fn notify_payment_path_successful(&self, path: &[&RouteHop]);
```

Comment:
Will `ChannelManager` need to care about these methods? Seems like the distinction isn't probing vs non-probing but rather routing vs event handling.

Reply:
`ChannelManager` as an async sending LSP has an opportunity to collect a lot of useful payment success/failure data for scoring. That said, I'm ok leaving these methods out for now if it's preferred

Reply:
Hmm... so essentially `ChannelManager` wouldn't surface any events for the user to handle but would handle them internally. Wouldn't it also need to retry failures? Seems like something more akin to `InvoicePayer` would be required rather than just the `Router`.

Reply:
Discussed offline, yes that's accurate but we're not sure what to do yet. W/r/t to the question at hand, would you prefer to remove these methods for now?

Reply:
> Discussed offline, yes that's accurate but we're not sure what to do yet. W/r/t to the question at hand, would you prefer to remove these methods for now?

Yeah, let's remove them... But that said, should we even bother with the interface churn now -- given this is for experimentation -- if it will likely to be different when accounting for retires? That is, can't the experiment simply call the standalone `find_route` function directly for now?

Reply:
I think we're actually adding trampoline v1 right now, which stands alone from the async payments experimentation that builds on top. cc @TheBlueMatt lmk if you have thoughts on this.

> if it will likely to be different when accounting for retires?

I don't see why the `Router` interface would be different when accounting for retries 🤔 (Aside from re-adding the `payment_path_*` methods for scoring)

Reply:
Right, i'm definitely gonna remove the `payment_path_*` methods for now. I was hoping to get your input on: 

> should we even bother with the interface churn now -- given this is for experimentation -- if it will likely to be different when accounting for retires? That is, can't the experiment simply call the standalone find_route function directly for now? 

@TheBlueMatt

Reply:
The concern is about essentially re-writing the retry logic of `InvoicePayer` here. I'm not yet convinced parameterizing `ChannelManager` with a `Router` is the right approach. Even if it is, it sorta begs the question as to why we shouldn't move all of `InvoicePayer`'s retry logic inside `ChannelManager` rather than using `InvoicePayer` as an `EventHandler` wrapper, as we currently do. Just think we need to put some more thought into this design.

Reply:
Discussed offline, gonna move forward with parameterizing `ChannelManager` with `Router` but reassess overall API design at each step. There may be an opportunity to break `ChannelManager` up into multiple subsystems (e.g. payment forwarding) to clean things up along the way. 

@jkczyz can speak more to this, but it may be advantageous to create a `Node` abstraction over `InvoicePayer` / `ChannelManager` (which could be renamed).